physics/discrete_trajectory_body.hpp

```diff
@@ -168,7 +208,26 @@ void DiscreteTrajectory<Frame>::ForgetAfter(Instant const& time) {
   // entry and all the entries that follow it.  This preserves any entry with
   // time == |time|.
   auto const it = timeline_.upper_bound(time);
+  if (downsampling_.has_value()) {
+    if (it != timeline_.end() &&
+        downsampling_->start_of_dense_timeline->first >= it->first) {
```

Comment:
This seems backwards, I think you want `<=`.  You are going to shorten the dense timeline so you need to adjust the number of dense intervals.  In the case covered by your predicate the dense timeline goes away entirely so it should be reset to empty.

If my analysis is correct, a test is missing.

Also, could we give name to the `first`?

Reply:
I think this is correct, PTAL with the new more verbose identifiers.
Updating the number of dense intervals is done systematically, *after* `erase` is called.
This simply ensures that in the case where the dense timeline goes away, its start is a valid iterator.

Reply:
(I will add a test tomorrow)